### PR TITLE
Added targetframeworks support to the duality package manager

### PR DESCRIPTION
**Pull request by Barsonax**
TODO:
- [x] Get nuget to work properly with packages that contain multiple targetframeworks
- [x] Add targetframeworks support to the unit test mockups
- [x] Add unit tests to test packages with multiple frameworks

Possible improvements:
- [ ] The targetframework (.NET 4.5.2) is hardcoded.
- [ ] Maybe there is a cleaner way for doing this. Open to suggestions.
- [ ] Update `Nuget.Core` to 2.14 to also support targets like netstandard (could be done in a separate PR though). In my test I found no breaking changes.

Changes:
- Decompiled the nuget PackageManager and copied it over to our solution. Cleaned up the class afterwards and added the code we need for targetframework functionality. I didnt see another way to do this due to shortcomings in the code of nuget. Its 250 lines but might be cleaned up a bit more as we don't use everything.

### CODE REVIEW

File: Source/Editor/DualityEditor/PackageManagement/DualityNugetPackageManager.cs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+			{
+				this.InstallPackage(package, this._targetFrameWork, ignoreDependencies, allowPrereleaseVersions, false);
+			}
+
+			public new void UpdatePackage(IPackage newPackage, bool updateDependencies, bool allowPrereleaseVersions)
+			{
+				this.Execute(newPackage, new UpdateWalker(this.LocalRepository, this.SourceRepository, new DependentsWalker(this.LocalRepository, this._targetFrameWork), NullConstraintProvider.Instance, this._targetFrameWork, this.Logger, updateDependencies, allowPrereleaseVersions));
+			}
+
+			private void Execute(IPackage package, IPackageOperationResolver resolver)
```
[the remark this answers is not in the document]
Comment (Barsonax) [reply Source/Editor/DualityEditor/PackageManagement/DualityNugetPackageManager.cs]: No changes where made but it was private and I needed that method

File: Source/Editor/DualityEditor/PackageManagement/DualityNugetPackageManager.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+			}
+
+			public override void InstallPackage(IPackage package, bool ignoreDependencies, bool allowPrereleaseVersions)
+			{
+				this.InstallPackage(package, this._targetFrameWork, ignoreDependencies, allowPrereleaseVersions, false);
+			}
+
+			public new void UpdatePackage(IPackage newPackage, bool updateDependencies, bool allowPrereleaseVersions)
+			{
+				this.Execute(newPackage, new UpdateWalker(this.LocalRepository, this.SourceRepository, new DependentsWalker(this.LocalRepository, this._targetFrameWork), NullConstraintProvider.Instance, this._targetFrameWork, this.Logger, updateDependencies, allowPrereleaseVersions));
```
[the remark this answers is not in the document]
Comment (Barsonax) [reply Source/Editor/DualityEditor/PackageManagement/DualityNugetPackageManager.cs]: Reason for existance of the class is in the comment on the class itself

File: Source/Editor/DualityEditor/PackageManagement/DualityNugetPackageManager.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+		{
+			this._packageManager.UninstallPackage(package, forceRemove, removeDependencies);
+		}
+
+		/// <summary>
+		/// Unfortunately nuget does not provide the ability to use a targetframework out of the box.
+		/// This is what this private class solves but due to some shortcomings in the nuget API it had to use some less desirable code such as hiding a method.
+		/// This is also why this class is private to prevent misuse.
+		/// </summary>
+		private class NugetFixedInternal : NuGet.PackageManager
```
[the remark this answers is not in the document]
Comment (Barsonax) [reply Source/Editor/DualityEditor/PackageManagement/DualityNugetPackageManager.cs]: done

File: Source/Editor/DualityEditor/PackageManagement/PackageManager.cs [hunk 4]
```diff
@@ -149,7 +149,8 @@ public PackageManager(PackageManagerEnvironment workEnvironment, PackageSetup pa
 				.Where(x => x != null)
 				.ToArray();
 			this.repository = new AggregateRepository(repositories);
-			this.manager = new NuGet.PackageManager(this.repository, this.env.RepositoryPath);
+			var targetFramework = new FrameworkName(".NETFramework", new Version(4, 5, 2));
```
[the remark this answers is not in the document]
Comment (Barsonax) [reply Source/Editor/DualityEditor/PackageManagement/PackageManager.cs]: Ah I have my style set to not use explicit if the type is obvious (which it is in this case as the type is on the right side)

File: Source/Editor/DualityEditor/PackageManagement/DualityNugetPackageManager.cs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+using NuGet;
+using NuGet.Resources;
+
+namespace Duality.Editor.PackageManagement
+{
+	/// <summary>
+	/// Provides nuget package functionality with targetframeworks.
+	/// Hides the hacks in <see cref="NugetFixedInternal"/> 
+	/// </summary>
+	public class DualityNugetPackageManager
```
[the remark this answers is not in the document]
Comment (Barsonax) [reply Source/Editor/DualityEditor/PackageManagement/DualityNugetPackageManager.cs]: Removed it

File: Source/Editor/DualityEditor/PackageManagement/DualityNugetPackageManager.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Barsonax) [reply Source/Editor/DualityEditor/PackageManagement/DualityNugetPackageManager.cs]: MockPackageSpec would need to be modified to support mocking multiple targetframeworks

**Comment by Barsonax** [comment]
When you install a package that triggers a update of another package nuget will internally use the update method. Since we didnt override this update method but hide it it uses the update method without the targetframework and thus it fails to correctly install the dependency.

I don't think this can be fixed. We would need to make a copy of the source of nuget to modify it to our needs.

**Comment by Barsonax** [comment]
Added some unit tests. Some things I noticed:
1. Nuget is still unable to properly parse .NetStandard targetframeworks from a nuspec. It will see it as `VersionUtility.UnsupportedFrameworkName` which is this `new FrameworkName("Unsupported", new Version())`.
2. Nuget is still unable to resolve a NetStandard dependency
3. Because of the above this fix does not enable duality to consume 'Netstandard' packages. It will make duality prefer the net452 target so a multi framework package that also targets net452 is now fine.

Possible fix for 1 and 2 might be updating nuget to atleast v2.12. This [page](https://www.nuget.org/packages/NETStandard.Library/) says that version should support netstandard. Maybe we won't even need the current modifications?

**Comment by ilexp** [comment]
Wow, wouldn't have expected that many additional issues popping up here. Glad to see you made it through regardless 👍

> So I now have arrived at a situation where all the unit tests are using the net452 (unless explicitly stated otherwise) target and they all complete succesfully. Next step is actually adding new tests to specifically test multi framework packages.

> Added some unit tests.

Means PR ready for the next review iteration?

> Possible fix for 1 and 2 might be updating nuget to atleast v2.12. This page says that version should support netstandard. Maybe we won't even need the current modifications?
> 
> EDIT: updated Nuget.Core to 2.14 (the latest version) as a test and now it does correctly parse netstandard as a target! With this change in theory all you have to do is change the targetframework you pass to our modified packagemanager to netstandard and it should then install the netstandard versions of the packages. We still need our custom NuGetTargetedPackageManager class though.

Great find. We should tackle an internal NuGet update to 2.14 in a separate PR.

**Comment by Barsonax** [comment]
> Wow, wouldn't have expected that many additional issues popping up here. Glad to see you made it through regardless 👍

Didn't expect the nuget library to have that many issues either. Not the best code I have seen, by a long shot.

> Means PR ready for the next review iteration?

Aye, this monday I will go on vacation for a month though so if there is anything left to do at that point it either has to wait or someone else (probably you :P) would have to do it.
 
> Great find. We should tackle an internal NuGet update to 2.14 in a separate PR.

Made the issue already: #711

**Comment by ilexp** [comment]
> Aye, this monday I will go on vacation for a month though so if there is anything left to do at that point it either has to wait or someone else (probably you :P) would have to do it.

Noted, thanks for the heads up. Probably not enough time from my side to fast-track this together with you this weekend, so I'll either pick this up in the mean time, or prepare the next review for when you return. Probably not that time critical for now.

> Made the issue already: #711

👍

**Comment by ilexp** [comment]
Started doing a few tweaks and fixes. As a side note, there are a lot of unrelated whitespace changes in this PR that make it hard to distinguish what actually changed - please try to avoid that kind of noise in future PRs.

One thing that remains an issue is unit test cases for .NET Standard target framework folders failing to set up properly, as .NET Standard is an unknown framework with the current NuGet version. This will be fixed in issue #711 - however, those unit tests need to remain operational until then. 

As the issue was introduced by changing MockPackageSpec to work on parsed NuGet `FrameworkName` instead of a `string` identifier directly, my approach here would be to revert that change and operate based on string identifiers again. They are converted to string when talking to NuGet anyway, and I think there's no reason to have this soft dependency in there, especially for unit test cases that might also intentionally use custom or unsupported framework folders which are by definition unknown to NuGet.

## Progress

- A few whitespace tweaks.
- Fixed Portable framework folder names in two cases where the conversion to `FrameworkName`-based mock specs accidentally switched them to .NET Framework 4.5.

## ToDo

- Revert back to `string`-based framework folders and identifiers in mock package specs, and update their new framework-dependent dependencies accordingly. Consider not using `DependencySet` for that, as that's not a test-specific class where validating framework names makes more sense.
- Consider removing `DependencySet`, if it is no longer used after the above test change.
- Consider simplifying `NuGetTargetedPackageManager` by removing unused parts, if any.
- Test real-world package install with the new changes.
- Review code again.

**Comment by Barsonax** [comment]
> As a side note, there are a lot of unrelated whitespace changes in this PR that make it hard to distinguish what actually changed - please try to avoid that kind of noise in future PRs.

Iam not sure why I get these whitespace changes. Was the file not normalized yet in git or is this some setting in my IDE that causes this?

**Comment by ilexp** [comment]
> Iam not sure why I get these whitespace changes. Was the file not normalized yet in git or is this some setting in my IDE that causes this?

Not sure how this happened, but if your IDE is doing this it should show up in the git diff before you commit. My guess would be some auto format setting?

**Comment by Barsonax** [comment]
Will have to check that when iam back home

**Comment by ilexp** [comment]
Tested self-install, package install and uninstall in a local setup, everything worked as expected. Tests are all green too, so no regressions expected. Merging this and triggering a binary release.

**Pull request merged**